### PR TITLE
chore: refactor `Client` initialization

### LINKED ISSUE #3748: fedimint-client Builder is error-prone
When attempting to add a [new function to read the federation ID](https://github.com/fedimint/fedimint/pull/3740/commits/d75e919069dea94493c2e2ac78e4174b40d624b0#diff-5c9c1d41afb86eeab67b808b3e245347ee5d1ae1e313ce358dc3c76e208fcb82R1497) from the client builder, CI tests failed because `get_config` seems to write the config to the DB. So when `get_config` is called again as part of `ClientBuilder::build()` it panicks since it wasn't expecting both a `FederationInfo` and a config stored in the DB.

`ClientBuilder` having an optional config as well as an optional db source feels quite ambiguous. Is there an invariant that must be satisfied? Must one of them be defined? Must the db source be `Reuse` is the FederationInfo is not there? Also why is the FederationInfo called `config`?

Discussed a bit with DPC and we might want to rethink this from a first principles approach.

### PR DESCRIPTION
**Pull request by dpc**
Fix #3748

* split the 3 cases of initalization (to be built on later)
* require only needed args, instead of using Option gymnastics

### CODE REVIEW

**Comment by dpc** [comment]
Need to add comments and fix restore test which will now require an empty db.

File: devimint/src/federation.rs [hunk 1]
```diff
@@ -36,41 +37,81 @@ pub struct Federation {
...
+
+        Ok(client)
+    }
+
+    /// Open or create a [`Client`] that starts with a fresh state.
+    pub async fn open_or_create(name: &str) -> Result<Client> {
+        let client = Self {
+            name: name.to_owned(),
+        };
+
```
Comment (douglaz) [inline devimint/src/federation.rs:82]: Missing `create_dir_all`?

**Review by douglaz — APPROVED** [review]
LGTM

**Comment by douglaz** [comment]
Should we backport this?

**Comment by dpc** [comment]
Added a tag, we'll see if it works and can decide then.

**Comment by fedimint-backports** [comment]
Backport failed for `releases/v0.2`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin releases/v0.2
git worktree add -d .worktree/backport-3918-to-releases/v0.2 origin/releases/v0.2
cd .worktree/backport-3918-to-releases/v0.2
git switch --create backport-3918-to-releases/v0.2
git cherry-pick -x 3746d51dc6d060ed8a10662a2ccdd065a501160e 742c84a71410e3fb4ab5f0170580f1047c1d2f93 e0a081e9234b5d2e4a418d16c2b026414a333379
```
